### PR TITLE
Allow constraints to include fixed parameters

### PR DESCRIPTION
**Pull request by a-r-williamson**
When generating samples from distributions with constraints imposed the constraints can't depend on any fixed parameters. Allowing for this will be useful for e.g. generating EM-bright populations.

### CODE REVIEW

File: pycbc/distributions/joint.py [hunk 1]
```diff
@@ -195,6 +199,59 @@ def _ensure_fieldarray(params):
...
+        n_samps : int
+            Length of each field or key, needed for populating values.
+
+        Returns
+        -------
+        samples : dict, FieldArray
+            Modified copy of input containing additional fields / keys.
+        """
+        for const in self._constraints:
+            sargs = get_constraint_static_params(const)
```
Comment (a-r-williamson) [inline pycbc/distributions/joint.py:238]:
```suggestion
            sargs = self.get_constraint_static_params(const)
```

**Comment by ahnitz** [comment]
@a-r-williamson Thanks for this! I see how this works and it makes me wonder if there is perhaps a faster way to implement this that does't require so much overhead. Would it make sense to instead modify the constraint class so that the constraint_arg essentially has any static parameters "baked in"? that would require less work I think than this and mean there'd be no slowdown from doing the extra evaluations for every sample drawn. It's not a big deal for simple things like drawing injection sets, but might be a bigger overhead for doing PE, for example (especially if using a fast likelihood method).

**Comment by ahnitz** [comment]
@a-r-williamson I am essentially thinking along the lines if this part here could be preprocessed (maybe internally by taking a static_params argument?) so that the string already has any replacements done for the static params.

**Comment by a-r-williamson** [comment]
Hi @ahnitz, totally agree about the issues this could cause for PE. I'm happy to switch this to an alternative approach along the lines you suggest.

File: bin/pycbc_create_injections [hunk 2]
```diff
@@ -205,7 +205,8 @@ dists = distributions.read_distributions_from_config(cp, opts.dist_section)
 
 # construct class that will draw the samples
 randomsampler = JointDistribution(variable_params, *dists,
-                               **{"constraints" : constraints})
+                               **{"constraints" : constraints,
+                                  "static_args" : static_params})
```
[the remark this answers is not in the document]
Comment (a-r-williamson) [reply bin/pycbc_create_injections]: This is not in the proposed changes anymore.

File: pycbc/distributions/joint.py [hunk 3]
```diff
@@ -195,6 +198,43 @@ def _ensure_fieldarray(params):
...
+        """
+        for const in self._constraints:
+            sargs = set([
+                s for s in get_fields_from_arg(const.constraint_arg)
+                if not s.isnumeric() and s in self.static_args.keys()])
+            if isinstance(samples, FieldArray):
+                for param in sargs:
+                    if param in samples.fieldnames:
+                        samples[param] = (
+                            numpy.ones(n_samps) * self.static_args[param])
```
[the remark this answers is not in the document]
Comment (a-r-williamson) [reply pycbc/distributions/joint.py]: And neither is this.

**Pull request merged**